backend/src/bundles/subscriptions/subscription.service.ts

```diff
@@ -0,0 +1,182 @@
+                status: HttpCode.INTERNAL_SERVER_ERROR,
+            });
+        }
+    }
+
+    public async updateCancelSubscribtion({
+        stripeSubscriptionId,
+        isCanceled,
+    }: CancelSubscriptionRequestDto): Promise<CancelSubscriptionResponseDto> {
+        if (!stripeSubscriptionId || typeof isCanceled !== 'boolean') {
+            throw new HttpError({
+                message:
+                    SubscriptionValidationMessage.SUBSCRIPTION_CANNOT_BE_CANCELED,
+                status: HttpCode.BAD_REQUEST,
+            });
+        }
+
+        try {
+            await stripeService.updateCancelSubscription({
+                stripeSubscriptionId,
+                isCanceled,
+            });
+
+            return { isCanceled };
+        } catch (error) {
+            throw new HttpError({
+                message: (error as Error).message,
+                status: HttpCode.INTERNAL_SERVER_ERROR,
+            });
+        }
+    }
+
+    public async webHookListener(body: unknown): Promise<void> {
+        const stripeResponse = body as Stripe.Event;
+        if (
+            !stripeResponse ||
+            !stripeResponse.type ||
+            !stripeResponse.data.object
+        ) {
+            return;
+        }
+
+        switch (stripeResponse.type) {
+            case SubscriptionWebHook.CUSTOMER_SUBSCRIPTION_UPDATED: {
+                const subscription = stripeResponse.data.object;
+                const updatedSubscription =
+                    await this.subscriptionRepository.updateByStripeSubscriptionId(
+                        subscription.id,
+                        {
+                            status: subscription.status,
+                            isCanceled: subscription.cancel_at_period_end,
+                            expiresAt: formatToDateFromUnix(
+                                subscription.current_period_end,
+                            ),
+                        },
+                    );
+
+                if (!updatedSubscription) {
+                    return;
+                }
+
+                const { userId } = updatedSubscription.toObject();
+
+                const activeSubscriptions =
+                    await this.subscriptionRepository.findAllActiveUserSubscriptions(
+                        userId,
+                    );
+
+                if (activeSubscriptions && activeSubscriptions.length > 1) {
+                    for (const [
+                        index,
+                        subscription,
+                    ] of activeSubscriptions.entries()) {
+                        if (index !== 0) {
+                            const { stripeSubscriptionId } =
+                                subscription.toNewObject();
+                            await stripeService.immediateCancelSubscription({
+                                stripeSubscriptionId,
+                            });
+                        }
+                    }
+                }
```

Comment:
why there is possibility that user will have more than one active subscription?

Reply:
When user changes his subscription, I create new subscription in stripe and the old subscription cancels.